README.md

```diff
@@ -26,21 +26,76 @@ pnpm add base-ts-result
 Interface that contains operation result and interaction methods
 ``` ts
 interface Result<Val, Err> {
+    // Contained Promise
     value: Val | Err;
-    isError: boolean;
 
+    // Queries
     unwrap(): Val;
-    unwrapOr(altRes: Val): Val;
     unwrapErr(): Err;
-
+    unwrapOr(altVal: Val): Val;
+    unwrapOrElse(altValFactory: (err: Err) => Val): Val;
     expect(msg: string): Val;
+    expectErr(msg: string): Err;
+    isOk(): this is OK<Val>;
+    isErr(): this is ERR<Err>;
+    ok(): Val|undefined,
+    err(): Err|undefined,
+
+    // Mappers
+    map<MappedVal>(mapper: (val: Val) => MappedVal): Result<MappedVal, Err>;
+    mapOrElse<MappedVal>(mapper: (val: Val) => MappedVal, fallback: (err: Err) => MappedVal): Result<MappedVal, Err>;
+    mapErr<MappedErr>(mapper: (err: Err) => MappedErr): Result<Val, MappedErr>;
+
+    // Utilities
+    inspect(inspector: (val: Val) => any): Result<Val, Err>;
+    inspectErr(inspector: (err: Err) => any): Result<Val, Err>;
+    toAsync(): AsyncResult<Val, Err>;
 }
 
 // Result implementations
 class OK implements Result;
 class ERR implements Result;
 ```
 
+### AsyncResult
+Class that contains operation result and interaction methods for async code
+``` ts
+type AsyncMapped<T> = T|Promise<T>
+// Async result implementation for more ergonomic usage of Results with async code
+class AsyncResult<Val, Err> {
+    // Contained promise
+    promise: Promise<Result<Val, Err>>; 
+
+    // Creators
+    static fromPromise<Val>(promise: Promise<Val>): AsyncResult<Val, unknown>;
```

Comment:
Can't specify error type here, better alternative:
```
static fromPromise<Val, Err = Error>( promise: Promise<Val>): AsyncResult<Val, Err> {
```

Reply:
Please, see my comment about type safety. I do not think user should be able to just specify error type, but narrow his error instead.

I was thinking of adding mapErr parameter to this function for convenience, but I decided against it because `AsyncResult.fromPromise(promise, () => 'err')` brings very little convenience over `AsyncResult.fromPromise(promise).mapErr(() => 'err')`, which offers more minimal API and IMO more clarity.

Reply:
fromPromise now returns `AsyncResult<Val, ResultCaughtError>`